File: src/test/java/conservative/DenominationConverterTest.java
package conservative;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Map;

public class DenominationConverterTest {

    private DenominationConverter converter;
    @BeforeEach
    public void setUp() {
        converter = new DenominationConverter();
    }

    @Test
    public void shouldConvert_USD_87pennies_intoExpectedAmount() {
        Map<Models.Denomination, Long> actualDenominations = converter.convert(87, "USD");
        Assertions.assertEquals(3, actualDenominations.get(Models.Dollar.QUARTER_COIN));
        Assertions.assertEquals(1, actualDenominations.get(Models.Dollar.DIME_COIN));
        Assertions.assertEquals(2, actualDenominations.get(Models.Dollar.PENNY_COIN));
    }

    @Test
    public void shouldConvert_USD_287pennies_intoExpectedAmount() {
        Map<Models.Denomination, Long> actualDenominations = converter.convert(287, "USD");
        Assertions.assertEquals(2, actualDenominations.get(Models.Dollar.ONE_DOLLAR_NOTE));
        Assertions.assertEquals(3, actualDenominations.get(Models.Dollar.QUARTER_COIN));
        Assertions.assertEquals(1, actualDenominations.get(Models.Dollar.DIME_COIN));
        Assertions.assertEquals(2, actualDenominations.get(Models.Dollar.PENNY_COIN));
    }

    @Test
    public void shouldConvert_EUR_287cent_intoExpectedAmount() {
        Map<Models.Denomination, Long> actualDenominations = converter.convert(287, "EUR");
        Assertions.assertEquals(1, actualDenominations.get(Models.Euro.TWO_EURO_COIN));
        Assertions.assertEquals(1, actualDenominations.get(Models.Euro.FIFTY_CENT_COIN));
        Assertions.assertEquals(1, actualDenominations.get(Models.Euro.TWENTY_CENT_COIN));
        Assertions.assertEquals(1, actualDenominations.get(Models.Euro.TEN_CENT_COIN));
        Assertions.assertEquals(1, actualDenominations.get(Models.Euro.FIVE_CENT_COIN));
        Assertions.assertEquals(1, actualDenominations.get(Models.Euro.TWO_CENT_COIN));
    }

    @Test
    public void shouldThrow_GBP() {
        Assertions.assertThrows(UnsupportedOperationException.class, () ->
            converter.convert(287, "GBP")
        );
    }
}

File: src/test/kotlin/progressive/DenominationConverterTest.kt
package progressive

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertEquals

class DenominationConverterTest {

    @Test
    fun `test for USD 87 pennies`() {
        val converter = DenominationConverter()
        val actualDenominations = converter.convert(87, "USD")
        assertEquals(3, actualDenominations[Dollar.QUARTER_COIN])
        assertEquals(1, actualDenominations[Dollar.DIME_COIN])
        assertEquals(2, actualDenominations[Dollar.PENNY_COIN])
        // assert no other keys
    }

    @Test
    fun `test for USD 287 pennies`() {
        val converter = DenominationConverter()
        val actualDenominations = converter.convert(287, "USD")
        assertEquals(2, actualDenominations[Dollar.ONE_DOLLAR_NOTE])
        assertEquals(3, actualDenominations[Dollar.QUARTER_COIN])
        assertEquals(1, actualDenominations[Dollar.DIME_COIN])
        assertEquals(2, actualDenominations[Dollar.PENNY_COIN])
        // assert no other keys
    }

    @Test
    fun `test for EUR 287 cent`() {
        val converter = DenominationConverter()
        val actualDenominations = converter.convert(287, "EUR")
        assertEquals(1, actualDenominations[Euro.TWO_EURO_COIN])
        assertEquals(1, actualDenominations[Euro.FIFTY_CENT_COIN])
        assertEquals(1, actualDenominations[Euro.TWENTY_CENT_COIN])
        assertEquals(1, actualDenominations[Euro.TEN_CENT_COIN])
        assertEquals(1, actualDenominations[Euro.FIVE_CENT_COIN])
        assertEquals(1, actualDenominations[Euro.TWO_CENT_COIN])
        // assert no other keys
    }

    @Test
    fun `test for GBP`() {
        val converter = DenominationConverter()
        assertThrows<UnsupportedOperationException> {
            converter.convert(287, "GBP")
        }
    }
}
